compose/ui/ui/src/desktopMain/kotlin/androidx/compose/ui/dnd/ExternalDrag.desktop.kt

```diff
@@ -0,0 +1,459 @@
+import androidx.compose.runtime.remember
+import androidx.compose.runtime.rememberUpdatedState
+import androidx.compose.runtime.setValue
+import androidx.compose.ui.ExperimentalComposeUiApi
+import androidx.compose.ui.Modifier
+import androidx.compose.ui.composed
+import androidx.compose.ui.geometry.Offset
+import androidx.compose.ui.geometry.Rect
+import androidx.compose.ui.graphics.painter.Painter
+import androidx.compose.ui.graphics.toPainter
+import androidx.compose.ui.layout.boundsInWindow
+import androidx.compose.ui.layout.onGloballyPositioned
+import androidx.compose.ui.window.LocalWindow
+import androidx.compose.ui.window.density
+import java.awt.Image
+import java.awt.Point
+import java.awt.Window
+import java.awt.datatransfer.DataFlavor
+import java.awt.datatransfer.DataFlavor.selectBestTextFlavor
+import java.awt.datatransfer.Transferable
+import java.awt.dnd.DnDConstants
+import java.awt.dnd.DropTarget
+import java.awt.dnd.DropTargetDragEvent
+import java.awt.dnd.DropTargetDropEvent
+import java.awt.dnd.DropTargetEvent
+import java.awt.dnd.DropTargetListener
+import java.awt.image.BufferedImage
+import java.io.File
+
+/**
+ * Represent data types drag and dropped to an application from outside.
+ */
+@ExperimentalComposeUiApi
+sealed interface DropData {
+    /**
+     * Represents list of files drag and dropped to an application in a raw [java.net.URI] format.
+     */
+    data class FilesList(val rawUris: List<String>) : DropData
+
+    /**
+     * Represents an image drag and dropped to an application.
+     */
+    data class Image(val painter: Painter) : DropData
+
+    /**
+     * Represent text drag and dropped to an application.
+     *
+     * @param mimeType mimeType of the [content] such as "text/plain", "text/html", etc.
+     */
+    data class Text(val content: String, val mimeType: String?) : DropData
+}
+
+/**
+ * Adds detector of external drag and drop (e.g. files DnD from Finder to an application)
+ *
+ * @param onDragStart will be called when the pointer with external content entered the component.
+ * @param onDrag will be called for all drag events inside the component.
+ * @param onDrop is called when the pointer is released with [DropData] the pointer held.
+ * @param onDragCancel is called if the pointer exited the component bounds or unknown data was dropped.
+ */
+@ExperimentalComposeUiApi
+@Composable
+fun Modifier.onExternalDrag(
+    enabled: Boolean = true,
+    onDragStart: (Offset) -> Unit = {},
+    onDrag: (Offset) -> Unit = {},
+    onDragCancel: () -> Unit = {},
+    onDrop: (DropData) -> Unit = {},
+): Modifier = composed {
+    if (!enabled) {
+        return@composed Modifier
+    }
+    val window = LocalWindow.current ?: return@composed Modifier
+
+    val onDragStartState by rememberUpdatedState(onDragStart)
+    val onDragState by rememberUpdatedState(onDrag)
+    val onDragCancelState by rememberUpdatedState(onDragCancel)
+    val onDropState by rememberUpdatedState(onDrop)
```

Comment:
That doesn't achieve the desired goal, because `installComponentDragHandler` stores the initial values of the functions and never updates them. You can fix it (and it even looks better, I think) like this: 
```
    val componentDragHandler = rememberUpdatedState(
        AwtWindowDropTarget.ComponentDragHandler(
            onDragStart = onDragStart,
            onDrag = onDrag,
            onDragCancel = onDragCancel,
            onDrop = onDrop
        )
    )
   ...
    private val handlers = mutableMapOf<Int, State<ComponentDragHandler>>()
   ...
    fun installComponentDragHandler(
        componentDragHandlerState: State<ComponentDragHandler>
    )
   ...
```

Reply:
Ah, you are right! Thank you! Will fix it